cpp/src/arrow/compute/kernels/scalar_nested_test.cc

```diff
@@ -116,6 +117,97 @@ TEST(TestScalarNested, ListElementInvalid) {
               Raises(StatusCode::Invalid));
 }
 
+void CheckListSlice(std::shared_ptr<Array> input, std::shared_ptr<Array> expected,
```

Comment:
A common thing that breaks kernels is to give the child array its own offset, in addition to the parent array; it'd be good to test that here.